# Comparing `tmp/ProtCHOIR-1.2.7-py3-none-any.whl.zip` & `tmp/ProtCHOIR-1.2.9-py3-none-any.whl.zip`

## zipinfo {}

```diff
@@ -1,20 +1,20 @@
-Zip file size: 92184 bytes, number of entries: 18
+Zip file size: 92241 bytes, number of entries: 18
 -rw-r--r--  2.0 unx    28505 b- defN 20-Jan-14 15:28 ProtCHOIR/AnalyseOligomer.py
--rw-r--r--  2.0 unx    46292 b- defN 20-Jan-14 15:28 ProtCHOIR/AnalyseProtomer.py
--rw-r--r--  2.0 unx    19156 b- defN 20-Jan-14 15:26 ProtCHOIR/Initialise.py
--rw-r--r--  2.0 unx    31114 b- defN 20-Jan-14 15:28 ProtCHOIR/MakeOligomer.py
+-rw-r--r--  2.0 unx    46720 b- defN 20-Jan-15 14:31 ProtCHOIR/AnalyseProtomer.py
+-rw-r--r--  2.0 unx    19442 b- defN 20-Jan-23 14:05 ProtCHOIR/Initialise.py
+-rw-r--r--  2.0 unx    31054 b- defN 20-Jan-15 14:05 ProtCHOIR/MakeOligomer.py
 -rw-r--r--  2.0 unx    54678 b- defN 20-Jan-14 15:28 ProtCHOIR/Toolbox.py
 -rw-r--r--  2.0 unx    25654 b- defN 20-Jan-14 15:26 ProtCHOIR/UpdateDatabases.py
 -rw-r--r--  2.0 unx        0 b- defN 18-Jun-04 09:10 ProtCHOIR/__init__.py
--rw-r--r--  2.0 unx    15034 b- defN 20-Jan-14 15:28 ProtCHOIR/__main__.py
+-rw-r--r--  2.0 unx    15026 b- defN 20-Jan-15 14:32 ProtCHOIR/__main__.py
 -rw-rw-r--  2.0 unx     5132 b- defN 19-Aug-29 15:16 ProtCHOIR/Contents/ProtCHOIR.svg
 -rw-rw-r--  2.0 unx     7348 b- defN 19-Aug-01 15:12 ProtCHOIR/Contents/ProtCHOIRReport.svg
 -rw-rw-r--  2.0 unx   181928 b- defN 19-Dec-12 14:10 ProtCHOIR/Contents/ProtCHOIRScheme.svg
 -rw-rw-r--  2.0 unx    13747 b- defN 19-Dec-10 13:45 ProtCHOIR/Contents/report_template.html
--rw-rw-r--  2.0 unx    18650 b- defN 20-Jan-14 15:28 ProtCHOIR-1.2.7.dist-info/LICENSE.txt
--rw-r--r--  2.0 unx     4842 b- defN 20-Jan-14 15:28 ProtCHOIR-1.2.7.dist-info/METADATA
--rw-r--r--  2.0 unx       92 b- defN 20-Jan-14 15:28 ProtCHOIR-1.2.7.dist-info/WHEEL
--rw-r--r--  2.0 unx       55 b- defN 20-Jan-14 15:28 ProtCHOIR-1.2.7.dist-info/entry_points.txt
--rw-r--r--  2.0 unx       10 b- defN 20-Jan-14 15:28 ProtCHOIR-1.2.7.dist-info/top_level.txt
-?rw-rw-r--  2.0 unx     1530 b- defN 20-Jan-14 15:28 ProtCHOIR-1.2.7.dist-info/RECORD
-18 files, 453767 bytes uncompressed, 89682 bytes compressed:  80.2%
+-rw-rw-r--  2.0 unx    18650 b- defN 20-Jan-23 14:07 ProtCHOIR-1.2.9.dist-info/LICENSE.txt
+-rw-r--r--  2.0 unx     4842 b- defN 20-Jan-23 14:07 ProtCHOIR-1.2.9.dist-info/METADATA
+-rw-r--r--  2.0 unx       92 b- defN 20-Jan-23 14:07 ProtCHOIR-1.2.9.dist-info/WHEEL
+-rw-r--r--  2.0 unx       55 b- defN 20-Jan-23 14:07 ProtCHOIR-1.2.9.dist-info/entry_points.txt
+-rw-r--r--  2.0 unx       10 b- defN 20-Jan-23 14:07 ProtCHOIR-1.2.9.dist-info/top_level.txt
+?rw-rw-r--  2.0 unx     1530 b- defN 20-Jan-23 14:07 ProtCHOIR-1.2.9.dist-info/RECORD
+18 files, 454413 bytes uncompressed, 89739 bytes compressed:  80.3%
```

## zipnote {}

```diff
@@ -30,26 +30,26 @@
 
 Filename: ProtCHOIR/Contents/ProtCHOIRScheme.svg
 Comment: 
 
 Filename: ProtCHOIR/Contents/report_template.html
 Comment: 
 
-Filename: ProtCHOIR-1.2.7.dist-info/LICENSE.txt
+Filename: ProtCHOIR-1.2.9.dist-info/LICENSE.txt
 Comment: 
 
-Filename: ProtCHOIR-1.2.7.dist-info/METADATA
+Filename: ProtCHOIR-1.2.9.dist-info/METADATA
 Comment: 
 
-Filename: ProtCHOIR-1.2.7.dist-info/WHEEL
+Filename: ProtCHOIR-1.2.9.dist-info/WHEEL
 Comment: 
 
-Filename: ProtCHOIR-1.2.7.dist-info/entry_points.txt
+Filename: ProtCHOIR-1.2.9.dist-info/entry_points.txt
 Comment: 
 
-Filename: ProtCHOIR-1.2.7.dist-info/top_level.txt
+Filename: ProtCHOIR-1.2.9.dist-info/top_level.txt
 Comment: 
 
-Filename: ProtCHOIR-1.2.7.dist-info/RECORD
+Filename: ProtCHOIR-1.2.9.dist-info/RECORD
 Comment: 
 
 Zip file comment:
```

## ProtCHOIR/AnalyseProtomer.py

```diff
@@ -784,15 +784,22 @@
             pdb_name = os.path.basename(fasta_file).split("_CHOIR_MonomerSequence.fasta")[0].replace('.', '_')
             records = list(SeqIO.parse(fasta_file, "fasta"))
             sequence = records[0].seq
             report['protomer_residues'] = str(len(sequence))
 
         elif args.input_file.endswith('.pdb'):
             report['input_filename'] = os.path.basename(input_file)
+            # Check number of chains in input and clean input file
             pdb_name, structure, nchains = pctools.parse_any_structure(input_file)
+            clean_input_file = os.path.join(workdir, pdb_name+'_Clean.pdb')
+            io.set_structure(structure)
+            io.save(clean_input_file, pctools.SelectIfCA())
+            # Reload clean file
+            pdb_name, structure, nchains = pctools.parse_any_structure(clean_input_file)
+            print(pdb_name)
             nchains, seqs, chain_ids = pctools.extract_seqs(structure, 0)
             sequence = seqs[0][1]
             report['protomer_residues'] = str(len(sequence))
             fasta_file = write_fasta(sequence)
             if nchains == 1:
                 print('Structure '+clrs['p']+pdb_name+clrs['n']+' is '+clrs['y']+'MONOMERIC'+clrs['n']+' as expected')
             else:
@@ -916,16 +923,16 @@
             report['hits'][hitchain]['id'] = str(round(hits[hitchain], 1))
             report['hits'][hitchain]['coverage'] = 'NA'
 
     pickle.dump(largest_oligo_complexes, open('CHOIR_OligoComplexes.pickle', 'wb'))
 
     if args.sequence_mode:
         if args.skip_conservation:
-            return (pdb_name, largest_oligo_complexes, interfaces_dict, tmdata), report, args
+            return (pdb_name, clean_input_file, largest_oligo_complexes, interfaces_dict, tmdata), report, args
         elif not args.skip_conservation:
-            return (pdb_name, largest_oligo_complexes, interfaces_dict, entropies, z_entropies, tmdata), report, args
+            return (pdb_name, clean_input_file, largest_oligo_complexes, interfaces_dict, entropies, z_entropies, tmdata), report, args
 
     elif not args.sequence_mode:
         if args.skip_conservation:
             return (pdb_name, clean_input_file, largest_oligo_complexes, interfaces_dict, residue_index_mapping, tmdata), report, args
         elif not args.skip_conservation:
             return (pdb_name, clean_input_file, largest_oligo_complexes, interfaces_dict, entropies, z_entropies, residue_index_mapping, minx, maxx, tmdata), report, args
```

## ProtCHOIR/Initialise.py

```diff
@@ -138,15 +138,15 @@
 def argument_parsing():
     parser = argparse.ArgumentParser(prog='ProtCHOIR',
                                      formatter_class=argparse.RawDescriptionHelpFormatter,
                                      description=description,
                                      epilog=epilogue)
 
     parser.add_argument('--version', action='version',
-                    version='%(prog)s 1.2.7')
+                    version='%(prog)s 1.2.9')
 
     parser.add_argument('-f', '--file',
                         dest='input_file',
                         type=str,
                         metavar='',
                         help='File containing the paths of pdbs to be oligomerized')
 
@@ -360,46 +360,52 @@
 
 
 maxasa = {'ALA':129.0,'ARG':274.0,'ASN':195.0,'ASP':193.0,
           'CYS':167.0,'GLU':223.0,'GLN':225.0,'GLY':104.0,
           'HIS':224.0,'ILE':197.0,'LEU':201.0,'LYS':236.0,
           'MET':224.0,'PHE':240.0,'PRO':159.0,'SER':155.0,
           'THR':172.0,'TRP':285.0,'TYR':263.0,'VAL':174.0,
-          'MME':224.0,'MSE':224.0}
+          'MME':224.0,'MSE':224.0, 'UNK':129.0}
 '''
 Theoretical maximum solvent accessible area values retrieved from:
 TIEN, M. Z., MEYER, A. G., SYDYKOVA, D. K., SPIELMAN, S. J., & WILKE, C. O.
 Maximum allowed solvent accessibilites of residues in proteins. PloS one, 2013.
+* B(ASX) takes the same value as D(ASP)
+* Z(GLX) takes the same value as E(GLU)
 '''
 
 #------------------------------------------------------------------------------
 hydro = {'ALA': -0.17, 'ARG': -0.81, 'ASN': -0.42, 'ASP': -1.23,
          'CYS': 0.24, 'GLN': -2.02, 'GLU': -0.58, 'GLY': -0.01,
          'HIS': -0.96, 'ILE': 0.31, 'LEU': 0.56, 'LYS': -0.99,
          'MET': 0.23, 'PHE': 1.13, 'PRO': -0.45, 'SER': -0.13,
          'THR': -0.14, 'TRP': 1.85, 'TYR': 0.94, 'VAL': -0.07,
-         'MME': 0.23, 'MSE': 0.23}
+         'MME': 0.23, 'MSE': 0.23, 'UNK': -0.17}
 '''
 Hydropobicity values retrieved from:
 WIMLEY, W. C.; WHITE, S. H. Experimentally determined hydrophobicity scale for
 proteins at membrane interfaces. Nature Structural Biology, 1996.
+* B(ASX) takes the same value as D(ASP)
+* Z(GLX) takes the same value as E(GLU)
 '''
 
 #------------------------------------------------------------------------------
 aa_bgf = {'A': 0.078, 'C': 0.024, 'D': 0.052, 'E': 0.059, 'F': 0.044,
           'G': 0.083, 'H': 0.025, 'I': 0.062, 'K': 0.056, 'L': 0.092,
           'M': 0.024, 'N': 0.041, 'P': 0.043, 'Q': 0.034, 'R': 0.051,
-          'S': 0.059, 'T': 0.055, 'V': 0.072, 'W': 0.014, 'Y': 0.034}
+          'S': 0.059, 'T': 0.055, 'V': 0.072, 'W': 0.014, 'Y': 0.034,
+          'B': 0.052, 'Z': 0.059}
 
 '''
 Amino acids background frequencies retrieved from:
 CAPRA, J. A.; SINGH, M. Predicting functionally important residues from sequence
 conservation. Bioinformatics, 2007.
+* B(ASX) takes the same value as D(ASP)
+* Z(GLX) takes the same value as E(GLU)
 '''
-True and False
 
 # Initialise
 ###############################################################################
 print(description)
 
 # Retrieve the paths from provided configuration file
 if (not initial_args.config_file or not os.path.isfile(initial_args.config_file)) and not os.path.isfile('CHOIR.cfg'):
```

## ProtCHOIR/MakeOligomer.py

```diff
@@ -140,19 +140,19 @@
         f.write("aln.append_model(mdl, align_codes='"+os.path.basename(renamed_chains_file)+"("+current_chain+")', atom_files='"+renamed_chains_file+"')\n")
         f.write("aln.append(file='"+g_input_file+"', align_codes='"+input_basename+"',alignment_format='FASTA', remove_gaps=False)\n")
         f.write('aln.align2d()\n')
         f.write("aln.write(file='"+temp_out+"', alignment_format='FASTA')")
     alignment_log = genali_file.replace('.py', '.log')
     spec = importlib.util.spec_from_file_location("genmodel", genali_file)
     genali_module = importlib.util.module_from_spec(spec)
-    #temp = sys.stdout
+    temp = sys.stdout
     sys.stdout = open(alignment_log, 'w')
     spec.loader.exec_module(genali_module)
-    sys.stdout.close()
-    #sys.stdout = temp
+    #sys.stdout.close()
+    sys.stdout = temp
     with open(temp_out, 'r') as infile, open(fasta_out, 'w') as outfile:
         n = 1
         for line in infile.readlines():
             if line.startswith('>'):
                 if n == 1:
                     outfile.write(line)
                 if n > 1:
@@ -621,15 +621,15 @@
         if args.multiprocess is True:
             p = Pool()
             for fasta_out, output in p.map_async(alignment_from_sequence, chains_dict.values()).get():
                 alignment_files.append(fasta_out)
                 print(output)
         else:
             for current_chain in chains_dict.values():
-                fasta_out, output = alignment_from_sequence(best_oligo_template_code, renamed_chains_file, input_file, current_chain)
+                fasta_out, output = alignment_from_sequence(current_chain)
                 alignment_files.append(fasta_out)
                 print(output)
     print('Alignment files:\n'+clrs['g']+('\n').join([os.path.basename(i) for i in alignment_files])+clrs['n'])
 
 
     # Generate final alignment which will be the input for Modeller
     final_alignment, full_residue_mapping = generate_ali(alignment_files, best_oligo_template_code, residue_index_mapping, args)
```

## ProtCHOIR/__main__.py

```diff
@@ -268,17 +268,17 @@
         elif analyse_protomer_results is not None and args.sequence_mode is True:
             residue_index_mapping = None
             minx = None
             maxx = None
             if args.skip_conservation:
                 entropies = None
                 z_entropies = None
-                pdb_name, largest_oligo_complexes, interfaces_dict, tmdata = analyse_protomer_results
+                pdb_name, clean_input_file, largest_oligo_complexes, interfaces_dict, tmdata = analyse_protomer_results
             elif not args.skip_conservation:
-                pdb_name, largest_oligo_complexes, interfaces_dict, entropies, z_entropies, tmdata = analyse_protomer_results
+                pdb_name, clean_input_file, largest_oligo_complexes, interfaces_dict, entropies, z_entropies, tmdata = analyse_protomer_results
                 if entropies == z_entropies == minx == maxx == None:
                     args.skip_conservation = True
 
 
         elif analyse_protomer_results is not None and args.sequence_mode is False:
             if args.skip_conservation:
                 minx = None
@@ -289,16 +289,15 @@
             elif not args.skip_conservation:
                 pdb_name, clean_input_file, largest_oligo_complexes, interfaces_dict, entropies, z_entropies, residue_index_mapping, minx, maxx, tmdata = analyse_protomer_results
                 if entropies == z_entropies == minx == maxx == None:
                     args.skip_conservation = True
 
         report['runtime_arguments']['skip_conservation'] = args.skip_conservation
 
-        if args.sequence_mode is False:
-            new_input_file = clean_input_file
+        new_input_file = clean_input_file
 
         # Use information of complexes to build oligomers
         best_oligo_template, built_oligomers, report = make_oligomer(new_input_file, largest_oligo_complexes, report, args, residue_index_mapping=residue_index_mapping)
 
         # If no models were built, exit nicely.
         if built_oligomers is None:
             finalize(report, input_basename, start_time, start_timestamp, args)
```

## Comparing `ProtCHOIR-1.2.7.dist-info/LICENSE.txt` & `ProtCHOIR-1.2.9.dist-info/LICENSE.txt`

 * *Files identical despite different names*

## Comparing `ProtCHOIR-1.2.7.dist-info/METADATA` & `ProtCHOIR-1.2.9.dist-info/METADATA`

 * *Files 0% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 Metadata-Version: 2.1
 Name: ProtCHOIR
-Version: 1.2.7
+Version: 1.2.9
 Summary: A Modeller-based pipeline to generate homo-oligomers.
 Home-page: https://github.com/monteirotorres/ProtCHOIR
 Author: Pedro Torres, Tom Blundell
 Author-email: phm34@cam.ac.uk.com
 License: CC-BY-4.0
 Keywords: PDB,structure,protein,oligomerization,oligomers,multimer
 Platform: UNKNOWN
```

## Comparing `ProtCHOIR-1.2.7.dist-info/RECORD` & `ProtCHOIR-1.2.9.dist-info/RECORD`

 * *Files 26% similar despite different names*

```diff
@@ -1,18 +1,18 @@
 ProtCHOIR/AnalyseOligomer.py,sha256=1GzgtndVFc6tLENguF6LGW3CdtV3cYFN1cgGFl5OgMQ,28505
-ProtCHOIR/AnalyseProtomer.py,sha256=fFDw89CVC6TsVYSEUHBB6Toky_3xSB349quY8kLPOqw,46292
-ProtCHOIR/Initialise.py,sha256=PTUiRhx6ISu6DVgzwKinKAsgXPJWvb79oZlfgkJ4WKU,19156
-ProtCHOIR/MakeOligomer.py,sha256=hrkk9UPF2lEiYx7HvXUtWmGosh1BfJCcczeckGUWQ_M,31114
+ProtCHOIR/AnalyseProtomer.py,sha256=7p9ZbQseBgtyde89Kkq_rSZqADb5GN1yzpnsx8U-Y68,46720
+ProtCHOIR/Initialise.py,sha256=NUoVdtwJUMQLODpDAuRkjybDFPzRQZKW3DjHxB9INps,19442
+ProtCHOIR/MakeOligomer.py,sha256=6AMxuPRvGRQ-5upg39C0BWm1RIorX941OTdbDb25A9g,31054
 ProtCHOIR/Toolbox.py,sha256=6231ilsF_IgNvBUae-I9tROWuEmqJJsD7GG2IEKknuc,54678
 ProtCHOIR/UpdateDatabases.py,sha256=C5P-eFok5P1v6C9XalyvThmWqQOJ4twadAFu9BetyJY,25654
 ProtCHOIR/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
-ProtCHOIR/__main__.py,sha256=1DjL-w4WNLbIqDRWCuxNjbNYdrwhjA3fU_DnxWnh4M0,15034
+ProtCHOIR/__main__.py,sha256=OOGAc8lj0XzI3Z5P3gZJy5AVTBrIGrQy-4GRPmi9vfk,15026
 ProtCHOIR/Contents/ProtCHOIR.svg,sha256=7956JDoHB31_OHHuwkfkBMuOk0uM92LT3uDzocKIopE,5132
 ProtCHOIR/Contents/ProtCHOIRReport.svg,sha256=57HXNJrCmgo2vGOB0wqbd9jha8nkAcVisuiSvlZX4ic,7348
 ProtCHOIR/Contents/ProtCHOIRScheme.svg,sha256=yVxhOffvAVGIqUU4sowhkMIkp-sTG4Jbn-AWWPEpn2M,181928
 ProtCHOIR/Contents/report_template.html,sha256=ttoQJODa5wRjs_pHZUaKZop0cvy1YKtS7rEqkg_zNok,13747
-ProtCHOIR-1.2.7.dist-info/LICENSE.txt,sha256=fnFw486_iKn2DHuEIUGDI8CTBNoa9NXpD02h3ByKJmE,18650
-ProtCHOIR-1.2.7.dist-info/METADATA,sha256=I3elVdAqd4SK2wjeSyKYww5cRfAUoIVHLVuiEz5WGvo,4842
-ProtCHOIR-1.2.7.dist-info/WHEEL,sha256=U88EhGIw8Sj2_phqajeu_EAi3RAo8-C6zV3REsWbWbs,92
-ProtCHOIR-1.2.7.dist-info/entry_points.txt,sha256=IrW_Xn_PuuyzJ_mwQnYVeco0aI_CUyWwtgTTfKHMhVw,55
-ProtCHOIR-1.2.7.dist-info/top_level.txt,sha256=0o2JLf271LmsuhjeanWXkwDJRSB_Tmynj2dRqnzuCT4,10
-ProtCHOIR-1.2.7.dist-info/RECORD,,
+ProtCHOIR-1.2.9.dist-info/LICENSE.txt,sha256=fnFw486_iKn2DHuEIUGDI8CTBNoa9NXpD02h3ByKJmE,18650
+ProtCHOIR-1.2.9.dist-info/METADATA,sha256=K9vTwnL5sCKO88NE-2BmNGtl4kw3b4nimfXyCLtsiis,4842
+ProtCHOIR-1.2.9.dist-info/WHEEL,sha256=U88EhGIw8Sj2_phqajeu_EAi3RAo8-C6zV3REsWbWbs,92
+ProtCHOIR-1.2.9.dist-info/entry_points.txt,sha256=IrW_Xn_PuuyzJ_mwQnYVeco0aI_CUyWwtgTTfKHMhVw,55
+ProtCHOIR-1.2.9.dist-info/top_level.txt,sha256=0o2JLf271LmsuhjeanWXkwDJRSB_Tmynj2dRqnzuCT4,10
+ProtCHOIR-1.2.9.dist-info/RECORD,,
```

